### ISSUE TITLE
[JENKINS-19099] Slow loading and large memory usage after 1.480.3 to 1.509.2 migration

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#14814]
<p>I used to run a Jenkins server with about 10 jobs with full history (~ 500 builds).</p>

<h5><a name="1.480.3"></a>1.480.3</h5>
<p><em>Server start -> wait for 1 min -> login -> dashboard</em><br/>
Loading time was decent (<1min) and memory usage is about 1GB.</p>

<h5><a name="1.509.2"></a>1.509.2</h5>
<p><em>Server start -> wait for 5 sec -> login -> wait for 10 min -> dashboard</em><br/>
After updating WAR to 1.509.2 and all plugins to latest version (like dashboard to 2.8), loading time is horrible (10min) and memory usage is about 4.5GB.</p>

<p>Navigation is faster, it seems that all data is in memory but I have to get back to 1.480.3 because such loading time and memory consumption is not acceptable on my jenkins server.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tarento">tarento</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099" target="_blank">Slow loading and large memory usage after 1.480.3 to 1.509.2 migration</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-07-24T07:11:47+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I used to run a Jenkins server with about 10 jobs with full history (~ 500 builds).</p>

<h5><a name="1.480.3"></a>1.480.3</h5>
<p><em>Server start -&gt; wait for 1 min -&gt; login -&gt; dashboard</em>
Loading time was decent (&lt;1min) and memory usage is about 1GB.</p>

<h5><a name="1.509.2"></a>1.509.2</h5>
<p><em>Server start -&gt; wait for 5 sec -&gt; login -&gt; wait for 10 min -&gt; dashboard</em>
After updating WAR to 1.509.2 and all plugins to latest version (like dashboard to 2.8), loading time is horrible (10min) and memory usage is about 4.5GB.</p>

<p>Navigation is faster, it seems that all data is in memory but I have to get back to 1.480.3 because such loading time and memory consumption is not acceptable on my jenkins server.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Debian Wheezy<br/>
OpenJDK 1.7b25<br/>
Tomcat 6
```
</details>
<details><summary><i>1 attachment</i></summary>

- [jenkins-threaddump.log](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/30228/jenkins-threaddump.log)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>kingluke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=188902&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-188902">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We are on 1.509.2 and also experiencing slow loading time on the initial entry page.   Once the initial load is complete, reloaded the page is fast.</p>
</pre>
</details>

</ul>
</details>

<p>We are on 1.509.2 and also experiencing slow loading time on the initial entry page.   Once the initial load is complete, reloaded the page is fast.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/seanw.jpg" title="seanw's avatar" /> seanw:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=189912&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-189912">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am on 1.539. I too have major memory usage. It quickly climbs to 5.5GB (the max I set it for) just after the service is started, without loading the UI. I tried disabling most plugins I thought were culprits.
Some plugins that I used (Fitnesse), I even temporarily disabled some calls like Parent.getBuilds(), which might have been the cause of loading all build history in memory. But there must be more places that load all the build history, instead of properly using the Lazy Loading of builds.</p>

<p>Any ideas?</p>

<p>Using Windows 7 with Java 64bit (7.45) with "Xrs -Xmx5g -XX:MaxPermSize=1g -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -jar "%BASE%\jenkins.war". (mainly as a test, definitely didn't expect all the 5GB of memory to quickly be used).</p>

<p>Plugins:
Name  ↓	Version   
analysis-core	1.54
ant	1.2
build-timeout	1.12
clone-workspace-scm	0.6
credentials	1.9.1	
cygpath	1.5
dry	2.37
email-ext	2.32	
fitnesse	1.9-SNAPSHOT (private-11/14/2013 23:29-Sean)
greenballs	1.12
join	1.15
ldap	1.2
mailer	1.5
matrix-auth	1.0.2
maven-plugin	2
msbuild	1.21
pam-auth	1
parameterized-trigger	2.2
schedule-failed-builds	1.1
ssh-credentials	1.5.1
ssh-slaves	1.5
subversion	1.5
thinBackup	1.7.3
token-macro	1.8.1
violations	0.7.11
ws-cleanup	0.16
xunit	1.5</p>
</pre>
</details>

</ul>
</details>

<p>I am on 1.539. I too have major memory usage. It quickly climbs to 5.5GB (the max I set it for) just after the service is started, without loading the UI. I tried disabling most plugins I thought were culprits.<br/>
Some plugins that I used (Fitnesse), I even temporarily disabled some calls like Parent.getBuilds(), which might have been the cause of loading all build history in memory. But there must be more places that load all the build history, instead of properly using the Lazy Loading of builds.</p>

<p>Any ideas?</p>

<p>Using Windows 7 with Java 64bit (7.45) with "Xrs -Xmx5g -XX:MaxPermSize=1g -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -jar "%BASE%\jenkins.war". (mainly as a test, definitely didn't expect all the 5GB of memory to quickly be used).</p>

<p>Plugins:<br/>
Name  ↓	Version   <br/>
analysis-core	1.54<br/>
ant	1.2<br/>
build-timeout	1.12<br/>
clone-workspace-scm	0.6<br/>
credentials	1.9.1	<br/>
cygpath	1.5<br/>
dry	2.37<br/>
email-ext	2.32	<br/>
fitnesse	1.9-SNAPSHOT (private-11/14/2013 23:29-Sean)<br/>
greenballs	1.12<br/>
join	1.15<br/>
ldap	1.2<br/>
mailer	1.5<br/>
matrix-auth	1.0.2<br/>
maven-plugin	2<br/>
msbuild	1.21<br/>
pam-auth	1<br/>
parameterized-trigger	2.2<br/>
schedule-failed-builds	1.1<br/>
ssh-credentials	1.5.1<br/>
ssh-slaves	1.5<br/>
subversion	1.5<br/>
thinBackup	1.7.3<br/>
token-macro	1.8.1<br/>
violations	0.7.11<br/>
ws-cleanup	0.16<br/>
xunit	1.5</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/seanw.jpg" title="seanw's avatar" /> seanw:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=190134&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-190134">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I put a bounty on this on FreedomSponcors if someone can get this done this year.
<a href="http://www.freedomsponsors.org/core/issue/408/slow-loading-and-large-memory-usage-after-14803-to-15092-migration" class="external-link" target="_blank" rel="nofollow noopener">http://www.freedomsponsors.org/core/issue/408/slow-loading-and-large-memory-usage-after-14803-to-15092-migration</a></p>
</pre>
</details>

</ul>
</details>

<p>I put a bounty on this on FreedomSponcors if someone can get this done this year.<br/>
<a href="http://www.freedomsponsors.org/core/issue/408/slow-loading-and-large-memory-usage-after-14803-to-15092-migration" class="external-link" target="_blank" rel="nofollow noopener">http://www.freedomsponsors.org/core/issue/408/slow-loading-and-large-memory-usage-after-14803-to-15092-migration</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=190293&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-190293">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I suppose you should provide memory dump at least.
Have you updated the core to 1.509.3/4? These versions contains many fixes for lazy classloader, which may help you.</p>
</pre>
</details>

</ul>
</details>

<p>I suppose you should provide memory dump at least.<br/>
Have you updated the core to 1.509.3/4? These versions contains many fixes for lazy classloader, which may help you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=209532&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-209532">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is the issue still this bad on recent Jenkins versions (no older than eight weeks)?</p>

<p>Note that due to lazy loading of build records, several latent bugs in plugins have been discovered. Also, any dashboard elements that need to load all builds of a project (e.g. how many builds are stable/unstable/failed) will result in longer loading times due to the basic design. The recommendation is to disable them, or if possible, limit to the last N builds, with smaller N being better.</p>
</pre>
</details>

</ul>
</details>

<p>Is the issue still this bad on recent Jenkins versions (no older than eight weeks)?</p>

<p>Note that due to lazy loading of build records, several latent bugs in plugins have been discovered. Also, any dashboard elements that need to load all builds of a project (e.g. how many builds are stable/unstable/failed) will result in longer loading times due to the basic design. The recommendation is to disable them, or if possible, limit to the last N builds, with smaller N being better.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=211313&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-211313">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No response to comment asking for updated information in three weeks, so resolving as Cannot Reproduce.</p>

<p>Given the age of this issue and versions involved, please file a new report if a similar issue occurs again on recent versions of Jenkins.</p>
</pre>
</details>

</ul>
</details>

<p>No response to comment asking for updated information in three weeks, so resolving as Cannot Reproduce.</p>

<p>Given the age of this issue and versions involved, please file a new report if a similar issue occurs again on recent versions of Jenkins.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cilap:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232957&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232957">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have the same issue.</p>

<p>I have a big history of the builds &gt; 500 builds and we could narrow it down to the mentioned upgrade.</p>

<p>So right now we have a jenkins configured with over 5GB HEAP and still it takes about 20 minutes to render the dashboard.</p>

<p>Afaik from my research the dashboard does not use any caching anymore. So this creates big issues on high load machines.
We have a SSD system with about 100.000 IOPS possible and the machine is taking ages.</p>

<p>Is it not better to use caching and to decide a user if he wants to wait for a forced (by user click) update of the cache? Or maybe to do it by a trigger in the times the jenkins is on low load e.g. in the night?
Is there a way to reenable the caching for the builds again?</p>

<p>Right now we cannot use jenkins and the dashboard anymore, since it is unusable slow. Hope someone can help us.
Here (attached file log) what we could find out via /threadDump URL of jenkins. Hope it helps</p>
</pre>
</details>

</ul>
</details>

<p>I have the same issue.</p>

<p>I have a big history of the builds > 500 builds and we could narrow it down to the mentioned upgrade.</p>

<p>So right now we have a jenkins configured with over 5GB HEAP and still it takes about 20 minutes to render the dashboard.</p>

<p>Afaik from my research the dashboard does not use any caching anymore. So this creates big issues on high load machines.<br/>
We have a SSD system with about 100.000 IOPS possible and the machine is taking ages.</p>

<p>Is it not better to use caching and to decide a user if he wants to wait for a forced (by user click) update of the cache? Or maybe to do it by a trigger in the times the jenkins is on low load e.g. in the night?<br/>
Is there a way to reenable the caching for the builds again?</p>

<p>Right now we cannot use jenkins and the dashboard anymore, since it is unusable slow. Hope someone can help us.<br/>
Here (attached file log) what we could find out via /threadDump URL of jenkins. Hope it helps</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cilap:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232958&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232958">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>the threaddump</p>
</pre>
</details>

</ul>
</details>

<p>the threaddump</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232959&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232959">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>@Darko,
Is it really related to this issue? I really doubt about it. Most probably you are using a newer Jenkins version.</p>

<p>It would be better to create another issue. Please also follow the guide: <a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>
</pre>
</details>

</ul>
</details>

<p>@​Darko,<br/>
Is it really related to this issue? I really doubt about it. Most probably you are using a newer Jenkins version.</p>

<p>It would be better to create another issue. Please also follow the guide: <a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232960&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232960">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>After a short dive into project-stats-plugin I conclude the issue is totally unrelated to Jenkins core. The plugin iterates through all builds and invokes their reload =&gt; LazyLoading gets overwhelmed. See <a href="https://github.com/jenkinsci/project-stats-plugin/blob/master/src/main/java/hudson/plugins/projectstats/NumBuildsStats.java#L48-61" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/project-stats-plugin/blob/master/src/main/java/hudson/plugins/projectstats/NumBuildsStats.java#L48-61</a></p>

<p>Please create a new issue to project-stats-plugin</p>
</pre>
</details>

</ul>
</details>

<p>After a short dive into project-stats-plugin I conclude the issue is totally unrelated to Jenkins core. The plugin iterates through all builds and invokes their reload => LazyLoading gets overwhelmed. See <a href="https://github.com/jenkinsci/project-stats-plugin/blob/master/src/main/java/hudson/plugins/projectstats/NumBuildsStats.java#L48-61" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/project-stats-plugin/blob/master/src/main/java/hudson/plugins/projectstats/NumBuildsStats.java#L48-61</a></p>

<p>Please create a new issue to project-stats-plugin</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232966&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232966">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yep, Project Statistics Plugin is basically unusable since 1.485. Get rid of it.</p>

<p>I added a note to its wiki page.</p>
</pre>
</details>

</ul>
</details>

<p>Yep, Project Statistics Plugin is basically unusable since 1.485. Get rid of it.</p>

<p>I added a note to its wiki page.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cilap:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-19099?focusedId=232981&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-232981">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>thank you for your support, created <a href="https://issues.jenkins.io/browse/JENKINS-29607" title="Dashboard loading takes ages" class="issue-link" data-issue-key="JENKINS-29607">JENKINS-29607</a></p>
</pre>
</details>

</ul>
</details>

<p>thank you for your support, created <a href="https://issues.jenkins.io/browse/JENKINS-29607" title="Dashboard loading takes ages" class="issue-link" data-issue-key="JENKINS-29607">JENKINS-29607</a></p>
